File: Soluciones_Ejercicios/java/ejercicio 4.java
// ============================================================
//  EJERCICIO 4 - Contar la cantidad de hojas del árbol
//  Una hoja es un nodo sin hijos izquierdo ni derecho.
// ============================================================

public class Main {

    static class Nodo {
        int dato;
        Nodo izquierda, derecha;
        Nodo(int valor) { this.dato = valor; }
    }

    // ── Funciones principales del ejercicio ─────────────────

    static int contarNodos(Nodo raiz) {
        if (raiz == null) return 0;
        return 1 + contarNodos(raiz.izquierda) + contarNodos(raiz.derecha);
    }

    static int contarHojas(Nodo raiz) {
        if (raiz == null) return 0;
        if (raiz.izquierda == null && raiz.derecha == null) return 1;
        return contarHojas(raiz.izquierda) + contarHojas(raiz.derecha);
    }

    static void mostrarHojas(Nodo raiz) {
        if (raiz == null) return;
        if (raiz.izquierda == null && raiz.derecha == null) {
            System.out.print(raiz.dato + " ");
            return;
        }
        mostrarHojas(raiz.izquierda);
        mostrarHojas(raiz.derecha);
    }

    // ── Visualización ────────────────────────────────────────

    static void imprimirArbol(Nodo nodo, String prefijo, boolean esIzquierda) {
        if (nodo == null) return;
        System.out.println(prefijo + (esIzquierda ? "|-->" : "\\-->") + " " + nodo.dato);
        String nuevo = prefijo + (esIzquierda ? "|    " : "     ");
        imprimirArbol(nodo.izquierda, nuevo, true);
        imprimirArbol(nodo.derecha,   nuevo, false);
    }

    static void imprimirArbolCompleto(Nodo raiz) {
        if (raiz == null) return;
        System.out.println(raiz.dato);
        imprimirArbol(raiz.izquierda, "", true);
        imprimirArbol(raiz.derecha,   "", false);
    }

    static void linea() { System.out.println("-".repeat(40)); }

    public static void main(String[] args) {
        Nodo raiz = new Nodo(10);
        raiz.izquierda       = new Nodo(5);
        raiz.derecha         = new Nodo(15);
        raiz.izquierda.izquierda = new Nodo(2);
        raiz.izquierda.derecha   = new Nodo(7);
        raiz.derecha.izquierda   = new Nodo(12);
        raiz.derecha.derecha     = new Nodo(20);
        raiz.izquierda.izquierda.izquierda = new Nodo(1);
        raiz.izquierda.izquierda.derecha   = new Nodo(3);
        raiz.derecha.derecha.izquierda     = new Nodo(18);
        raiz.derecha.derecha.derecha       = new Nodo(25);

        linea();
        System.out.println("      --- EJERCICIO 4 (Java) ---");
        System.out.println("        Contar hojas del arbol");
        linea();

        System.out.println("\nArbol Estructurado:");
        imprimirArbolCompleto(raiz);

        linea();
        System.out.println("Nodos totales:     " + contarNodos(raiz));
        System.out.println("Cantidad de hojas: " + contarHojas(raiz));
        System.out.print("Nodos hoja:        "); mostrarHojas(raiz); System.out.println();
        linea();
    }
}

File: Soluciones_Ejercicios/c++/ejercicio4.cpp
#include <iostream>
#include <string>
using namespace std;

// ============================================================
//  EJERCICIO 4 - Contar la cantidad de hojas del árbol
//  Una hoja es un nodo sin hijos (izquierda == nullptr
//  Y derecha == nullptr al mismo tiempo).
// ============================================================

struct Nodo {
    int dato;
    Nodo* izquierda;
    Nodo* derecha;
    Nodo(int valor) : dato(valor), izquierda(nullptr), derecha(nullptr) {}
};

// ── Función principal del ejercicio ─────────────────────────

int contarHojas(Nodo* raiz) {
    if (!raiz) return 0;
    // Si no tiene hijos, es una hoja
    if (!raiz->izquierda && !raiz->derecha) return 1;
    return contarHojas(raiz->izquierda) + contarHojas(raiz->derecha);
}

int contarNodos(Nodo* raiz) {
    if (!raiz) return 0;
    return 1 + contarNodos(raiz->izquierda) + contarNodos(raiz->derecha);
}

// ── Mostrar hojas ────────────────────────────────────────────

void mostrarHojas(Nodo* raiz) {
    if (!raiz) return;
    if (!raiz->izquierda && !raiz->derecha) {
        cout << raiz->dato << " ";
        return;
    }
    mostrarHojas(raiz->izquierda);
    mostrarHojas(raiz->derecha);
}

// ── Visualización del árbol ─────────────────────────────────

void imprimirArbol(Nodo* raiz, string prefijo = "", bool esIzquierda = true) {
    if (!raiz) return;
    cout << prefijo << (esIzquierda ? "|-->" : "\\-->") << " " << raiz->dato << "\n";
    string nuevo = prefijo + (esIzquierda ? "|    " : "     ");
    imprimirArbol(raiz->izquierda, nuevo, true);
    imprimirArbol(raiz->derecha,   nuevo, false);
}

void imprimirArbolCompleto(Nodo* raiz) {
    if (!raiz) return;
    cout << raiz->dato << "\n";
    imprimirArbol(raiz->izquierda, "", true);
    imprimirArbol(raiz->derecha,   "", false);
}

void linea() { cout << string(40, '-') << "\n"; }

int main() {
    // Árbol del ejercicio 2 (11 nodos)
    Nodo* raiz = new Nodo(10);
    raiz->izquierda            = new Nodo(5);
    raiz->derecha              = new Nodo(15);
    raiz->izquierda->izquierda = new Nodo(2);
    raiz->izquierda->derecha   = new Nodo(7);
    raiz->derecha->izquierda   = new Nodo(12);
    raiz->derecha->derecha     = new Nodo(20);
    raiz->izquierda->izquierda->izquierda = new Nodo(1);
    raiz->izquierda->izquierda->derecha   = new Nodo(3);
    raiz->derecha->derecha->izquierda     = new Nodo(18);
    raiz->derecha->derecha->derecha       = new Nodo(25);

    linea();
    cout << "         --- EJERCICIO 4 ---\n";
    cout << "       Contar hojas del arbol\n";
    linea();

    cout << "\nArbol Estructurado:\n";
    imprimirArbolCompleto(raiz);

    linea();
    cout << "Nodos totales:   " << contarNodos(raiz) << "\n";
    cout << "Cantidad de hojas: " << contarHojas(raiz) << "\n";
    cout << "Nodos hoja:      "; mostrarHojas(raiz); cout << "\n";
    linea();

    return 0;
}
